content/blog/2019-11-08-unified-theory-gc/index.md

```diff
@@ -0,0 +1,245 @@
+object should not be freed if it could be used in the future. A conservative
+approximation of this, without any program analysis, is that an object could be
+used in the future if there exists a path of pointers to the object which
+originates from the stack or from a register. We call the starting points of
+such paths (i.e., all objects to which there is a direct pointer on the stack or
+in a register) the roots of the graph, which make up the multiset $R$.
+
+Using these definitions, we can formulate the reference counts of objects
+(denoted $\rho(v)$ for $v \in V$) as a fixed point of the following equation:
+
+$$ \rho(v) = \big|[v : v \in R]\big| +
+             \big|[(w, v) : (w, v) \in E \land \rho(w) > 0]\big| $$
+
+Here we recursively define the reference count of an object $v$ to be the number
+of root pointers to $v$ plus the number of pointers to $v$ from objects which
+themselves have non-zero reference counts. Any object whose reference count is
+zero according a fixed point of $\rho$ can be freed, as there is no way for the
+program to reference it in the future.
+
+However, it is important to note that there could be multiple fixed points to
+this equation, namely in the presence of cyclic garbage. If object $a$ points to
+object $b$, and vice versa, but neither is a root and neither is pointed to from
+elsewhere, then, by this formulation, both $\rho(a) = \rho(b) = 1$ and $\rho(a)
+= \rho(b) = 0$ are valid solutions. Ideally, a garbage collection algorithm will
+find the least fixed point of $\rho$, meaning that it will consider all cyclic
+garbage as able to be freed. A tracing collector does this, whereas a reference
+counting collector does not detect any cyclic garbage, and thus finds the
+greatest fixed point.
+
+### Alignments of Algorithmic Structures
+
+As modeled in the paper, a tracing garbage collector fundamentally works by
+performing a traversal on the memory graph, starting from the root objects in
+$R$. Every time it reaches an object, it increments its reference count. And,
+when it reaches an object for the first time, it traverses the pointers from
+that object. In order to complete the traversal, the algorithm maintains a work
+list, $W$, of objects that it needs to visit. This traversal recomputes all
+reference counts completely each time it runs. When it terminates, every live
+object will have a positive reference count, and every dead object will have a
+reference count of zero.
+
+This formulation of the algorithm differs in a few ways from a standard tracing
+collector. For example, there is no need in a real tracing collector to keep
+track of the full reference count of each object; it is sufficient to keep just
+a single mark bit which determines whether the reference count is non-zero. A
+tracing collector might also employ other optimizations, such as pointer
+reversal, to improve its run time or memory usage. However, these optimizations
```

Comment:
Pointer reversal is sort of an advanced topic we haven't talked about here—maybe just link to a basic explanation somewhere?